src/QsCompiler/CompilerOptimizations/Evaluation.fs

```diff
@@ -13,170 +13,181 @@ open Microsoft.Quantum.QsCompiler.SyntaxTree
 type private FunctionInterrupt =
 /// Represents the function invoking a return statement
 | Returned of TypedExpression
 /// Represents the function invoking a fail statement
 | Failed of TypedExpression
+/// We reached the limit of statement executions
+| OutOfStatements
 /// We were unable to evaluate the function for some reason
 | CouldNotEvaluate of string
 
-/// The current state of a function evaluation.
-/// The Ok case means that we are evaluating the function, inside normal control flow.
-/// The Ok case stores a map from all local variables to their current values.
-/// The Error case means that we are outside normal control flow, in some kind of interrupt.
-/// The Error case stores a FunctionInterrupt that describes the cause of the interrupt.
-/// I use the Result type to match the common design pattern of railway-oreiented programming.
-type private FunctionState = Result<Constants<TypedExpression>, FunctionInterrupt>
+/// A shorthand for the specific Imperative type used by several functions in this file
+type private Imp<'t> = Imperative<EvalState, 't, FunctionInterrupt>
+
+/// Represents a computation that decreases the remaining statements counter by 1.
+/// Yields an OutOfStatements interrupt if this decreases the remaining statements below 0.
+let private incrementState: Imp<Unit> = imperative {
+    let! vars, counter = getState
+    if counter < 1 then yield OutOfStatements
+    do! putState (vars, counter - 1)
+}
+
+/// Represents a computation that sets the given variables to the given values
+let private setVars callables entry: Imp<Unit> = imperative {
+    let! vars, counter = getState
+    do! putState (defineVarTuple (isLiteral callables) vars entry, counter)
+}
 
 
 /// Evaluates functions by stepping through their code
-type internal FunctionEvaluator(callables: Callables, maxRecursiveDepth: int) =
+type internal FunctionEvaluator(callables: Callables) =
 
-    /// Transforms a BoolLiteral into the corresponding bool
-    let castToBool f x =
+    /// Casts a BoolLiteral to the corresponding bool
+    let castToBool x: Imp<bool> = imperative {
         match x.Expression with
-        | BoolLiteral b -> Ok (f b)
-        | _ -> "Not a BoolLiteral: " + (printExpr x.Expression) |> CouldNotEvaluate |> Error
-
-    /// The callback for the subroutine that evaluates and simplifies an expression
-    member internal this.evaluateExpression constants expr =
-        ExpressionEvaluator(callables, constants, maxRecursiveDepth - 1).Transform expr
+        | BoolLiteral b -> return b
+        | _ -> yield CouldNotEvaluate ("Not a BoolLiteral: " + (printExpr x.Expression))
+    }
+
+    /// Evaluates and simplifies a single Q# expression
+    member internal __.evaluateExpression expr: Imp<TypedExpression> = imperative {
+        let! vars, counter = getState
+        let result = ExpressionEvaluator(callables, vars, counter / 2).Transform expr
```

Comment:
What's the reason for the counter / 2 ?

Reply:
This is a workaround for limiting the number of statements evaluated in recursive functions. Currently, evaluting a recursive call means going through the ExpressionEvaluator, which doesn't use the Imperative monad, so after trying to evaluate a recursive call we have no information about how many statements were executed during this call. This leads to edge cases where any given stack includes a small number of statement executions, but the total number of evaluations is exponentially large. Dividing by 2 blocks almost all of these edge cases, so we don't use up a ton of time trying to evaluate code. The "good" solution would require tracking the number of statements evaluated in the ExpressionEvaluator, which would require completely rewriting the ExpressionEvaluator to use the Imperative monad.

Reply:
Ok, seems fine to leave it like that. Thanks for the explanation!